Cannot load PDF when there is a # in the project directory name

Hi,
I installed your plugin in Pycharm (2021.1.2 RC) but all pdf documents open as a blank document.
My iOS is MacOS Big Sur 11.3.1.
Thank you.

I tried it and I got a fatal error. I also sent a report to author.

com.intellij.diagnostic.PluginException:  [Plugin: com.firsttimeinforever.intellij.pdf.viewer.intellij-pdf-viewer]
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.instantiate(ActionManagerImpl.java:250)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.convertStub(ActionManagerImpl.java:220)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.getActionImpl(ActionManagerImpl.java:551)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.preloadActions(ActionManagerImpl.java:1601)
	at com.intellij.openapi.actionSystem.impl.ActionPreloader.preload(ActionPreloader.java:14)
	at com.intellij.openapi.application.Preloader.lambda$preload$0(Preloader.java:84)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:688)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:634)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:64)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.openapi.application.Preloader.lambda$preload$1(Preloader.java:74)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:216)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:27)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:195)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:213)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:184)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.instantiate(ActionManagerImpl.java:240)
	... 23 more
Caused by: java.lang.NoClassDefFoundError: com/firsttimeinforever/intellij/pdf/viewer/model/HistoryNavigationDirection
	at com.firsttimeinforever.intellij.pdf.viewer.actions.navigation.PdfNavigateByHistoryAction$Forward.<init>(PdfNavigateByHistoryAction.kt:14)
	... 28 more
Caused by: java.lang.ClassNotFoundException: com.firsttimeinforever.intellij.pdf.viewer.model.HistoryNavigationDirection PluginClassLoader(plugin=PluginDescriptor(name=PDF Viewer, id=com.firsttimeinforever.intellij.pdf.viewer.intellij-pdf-viewer, descriptorPath=plugin.xml, path=~/Library/Application Support/JetBrains/PyCharm2021.1/plugins/plugin-0.12.0-alpha.1.jar, version=0.12.0-alpha.1, package=null), packagePrefix=null, instanceId=5, state=active)
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClass(PluginClassLoader.java:215)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 29 more

--- a/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt
+++ b/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt
@@ -3,26 +3,33 @@ package com.firsttimeinforever.intellij.pdf.viewer.jcef
 import com.firsttimeinforever.intellij.pdf.viewer.settings.PdfViewerSettings
 import com.firsttimeinforever.intellij.pdf.viewer.utility.PdfResourceLoader
 import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.util.Url
 import com.intellij.util.Urls
 import com.intellij.util.io.URLUtil
 import io.netty.buffer.Unpooled
 import io.netty.channel.ChannelHandlerContext
 import io.netty.handler.codec.http.FullHttpRequest
+import io.netty.handler.codec.http.HttpHeaderNames
+import io.netty.handler.codec.http.HttpMethod
+import io.netty.handler.codec.http.HttpResponseStatus
+import io.netty.handler.codec.http.HttpUtil
 import io.netty.handler.codec.http.QueryStringDecoder
+import io.netty.handler.stream.ChunkedStream
 import org.jetbrains.ide.BuiltInServerManager
 import org.jetbrains.ide.HttpRequestHandler
 import org.jetbrains.io.FileResponses
+import org.jetbrains.io.addKeepAliveIfNeeded
+import org.jetbrains.io.flushChunkedResponse
 import org.jetbrains.io.response
 import org.jetbrains.io.send
-import java.nio.file.Path
 import java.nio.file.Paths
-import java.util.*
-import kotlin.io.path.extension
 import kotlin.random.Random
 
 internal class PdfStaticServer : HttpRequestHandler() {
   private val serverUrl = "http://localhost:${BuiltInServerManager.getInstance().port}/$uuid"
+
+  private val vfsMap = mutableMapOf<String, VirtualFile>()
 
   init {
     logger.debug("Starting static server with url: $serverUrl")
@@ -53,15 +60,32 @@ internal class PdfStaticServer : HttpRequestHandler() {
     return path.startsWith("/get-file/")
   }
 
-  private fun assertValidExternalPath(path: Path) {
-    check(path.extension.lowercase(Locale.getDefault()) == "pdf") { "Only pdf files can be served from the outside of jar!" }
-  }
-
-  private fun sendExternalFile(path: String, context: ChannelHandlerContext, request: FullHttpRequest) {
-    val targetFile = Paths.get(path)
-    assertValidExternalPath(targetFile)
-    logger.debug("Sending external file: $targetFile")
-    FileResponses.sendFile(request, context.channel(), Paths.get(targetFile.toString()))
+  private fun sendExternalFile(url: String, context: ChannelHandlerContext, request: FullHttpRequest) {
+    val file: VirtualFile? = vfsMap[url]
+    val channel = context.channel()
+    if (file == null) {
+      logger.debug("Cannot find pdf file $url")
+      HttpResponseStatus.NOT_FOUND.send(channel, request)
+    } else if (file.isInLocalFileSystem) {
+      val path = Paths.get(file.path)
+      logger.debug("Sending external file, path: $path")
+      FileResponses.sendFile(request, channel, path)
+    } else {
+      logger.debug("Sending external file, url: $url")
+      // see org.jetbrains.builtInWebServer.StaticFileHandler
+      val response = FileResponses.prepareSend(request, channel, file.timeStamp, file.name) ?: return
+      val isKeepAlive = response.addKeepAliveIfNeeded(request)
+      // remove accept-ranges as we don't support
+      response.headers().remove(HttpHeaderNames.ACCEPT_RANGES)
+      if (request.method() != HttpMethod.HEAD) {
+        HttpUtil.setContentLength(response, file.length)
+      }
+      channel.write(response)
+      if (request.method() != HttpMethod.HEAD) {
+        channel.write(ChunkedStream(file.inputStream))
+      }
+      flushChunkedResponse(channel, isKeepAlive)
+    }
   }
 
   private fun makeInternalPath(path: String): String {
@@ -85,14 +109,19 @@ internal class PdfStaticServer : HttpRequestHandler() {
     response.send(context.channel(), request)
   }
 
-  fun getPreviewUrl(filePath: String, withReloadSalt: Boolean = false): String {
+  fun getPreviewUrl(file: VirtualFile, withReloadSalt: Boolean = false): String {
     val salt = if (withReloadSalt) Random.nextInt() else 0
+    vfsMap[file.url] = file
     val url = parseEncodedPath("$serverUrl/index.html")
     val server = BuiltInServerManager.getInstance()
     return server.addAuthToken(url)
       // `file` would be read via `urlDecoder.path()`, which calls `decodeComponent`
-      .addParameters(mapOf("__reloadSalt" to "$salt", "file" to URLUtil.encodeURIComponent("get-file/$filePath")))
+      .addParameters(mapOf("__reloadSalt" to "$salt", "file" to URLUtil.encodeURIComponent("get-file/${file.url}")))
       .toExternalForm()
+  }
+
+  fun disposePreviewUrl(file: VirtualFile) {
+    vfsMap.remove(file.url)
   }
 
   private fun parseEncodedPath(target: String): Url {

--- a/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt
+++ b/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt
@@ -5,6 +5,7 @@ import com.firsttimeinforever.intellij.pdf.viewer.utility.PdfResourceLoader
 import com.intellij.openapi.diagnostic.logger
 import com.intellij.util.Url
 import com.intellij.util.Urls
+import com.intellij.util.io.URLUtil
 import io.netty.buffer.Unpooled
 import io.netty.channel.ChannelHandlerContext
 import io.netty.handler.codec.http.FullHttpRequest
@@ -86,9 +87,12 @@ internal class PdfStaticServer : HttpRequestHandler() {
 
   fun getPreviewUrl(filePath: String, withReloadSalt: Boolean = false): String {
     val salt = if (withReloadSalt) Random.nextInt() else 0
-    val url = parseEncodedPath("$serverUrl/index.html?__reloadSalt=$salt&file=get-file/$filePath")
+    val url = parseEncodedPath("$serverUrl/index.html")
     val server = BuiltInServerManager.getInstance()
-    return server.addAuthToken(url).toExternalForm()
+    return server.addAuthToken(url)
+      // `file` would be read via `urlDecoder.path()`, which calls `decodeComponent`
+      .addParameters(mapOf("__reloadSalt" to "$salt", "file" to URLUtil.encodeURIComponent("get-file/$filePath")))
+      .toExternalForm()
   }
 
   private fun parseEncodedPath(target: String): Url {

--- a/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/ui/editor/view/PdfJcefPreviewController.kt
+++ b/plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/ui/editor/view/PdfJcefPreviewController.kt
@@ -129,7 +129,7 @@ class PdfJcefPreviewController(val project: Project, val virtualFile: VirtualFil
   private fun doActualReload(tryToPreserveState: Boolean = false) {
     viewLoaded = false
     try {
-      val base = PdfStaticServer.instance.getPreviewUrl(virtualFile.path, withReloadSalt = true)
+      val base = PdfStaticServer.instance.getPreviewUrl(virtualFile, withReloadSalt = true)
       if (firstLoad) {
         viewState = viewState.copy(sidebarViewMode = PdfViewerSettings.instance.defaultSidebarViewMode)
       }
@@ -255,7 +255,10 @@ class PdfJcefPreviewController(val project: Project, val virtualFile: VirtualFil
     pipe.send(IdeMessages.NavigateHistory(direction))
   }
 
-  override fun dispose() = Unit
+  override fun dispose() {
+    logger.debug("dispose $virtualFile")
+    PdfStaticServer.instance.disposePreviewUrl(virtualFile)
+  }
 
   companion object {
     private val logger = logger<PdfJcefPreviewController>()

